pkg/discovery/discoverymanager.go

```diff
@@ -0,0 +1,321 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	failedConfigs = prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "parca_agent_sd_failed_configs",
+			Help: "Current number of service discovery configurations that failed to load.",
+		},
+	)
+	discoveredTargets = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "parca_agent_sd_discovered_targets",
+			Help: "Current number of discovered targets.",
+		},
+		[]string{"config"},
+	)
+	receivedUpdates = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "parca_agent_sd_received_updates_total",
+			Help: "Total number of update events received from the SD providers.",
+		},
+	)
+	delayedUpdates = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "parca_agent_sd_updates_delayed_total",
+			Help: "Total number of update events that couldn't be sent immediately.",
+		},
+	)
+	sentUpdates = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "parca_agent_sd_updates_total",
+			Help: "Total number of update events sent to the SD consumers.",
+		},
+	)
```

Comment:
Is it better to make these metrics non-global? Maybe better to put them under a struct and with promauto.

Reply:
Thanks for the suggestion! Made a [PR](https://github.com/parca-dev/parca-agent/pull/182) for the same.